src/app/3d/qgs3dmaptoolmeasureline.cpp

```diff
@@ -0,0 +1,304 @@
+/***************************************************************************
+  qgs3dmaptoolmeasureline.cpp
+  --------------------------------------
+  Date                 : Jun 2019
+  Copyright            : (C) 2019 by Ismail Sunni
+  Email                : imajimatika at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include <QDebug>
+
+#include <Qt3DRender/QObjectPicker>
+#include <Qt3DRender/QPickEvent>
+#include <QKeyEvent>
+
+#include "qgsterrainentity_p.h"
+#include "qgs3dmaptoolmeasureline.h"
+#include "qgs3dutils.h"
+#include "qgs3dmapscene.h"
+#include "qgs3dmapcanvas.h"
+#include "qgsvectorlayer.h"
+#include "qgslinestring.h"
+#include "qgspoint.h"
+#include "qgsfeature.h"
+#include "qgsline3dsymbol.h"
+#include "qgsvectorlayer3drenderer.h"
+#include "qgsmaplayer.h"
+#include "qgs3dmeasuredialog.h"
+
+#include "qgs3dmapscenepickhandler.h"
+
+class Qgs3DMapToolMeasureLinePickHandler : public Qgs3DMapScenePickHandler
+{
+  public:
+    Qgs3DMapToolMeasureLinePickHandler( Qgs3DMapToolMeasureLine *measureLineTool ): mMeasureLineTool( measureLineTool ) {}
+    void handlePickOnVectorLayer( QgsVectorLayer *vlayer, QgsFeatureId id, const QVector3D &worldIntersection, Qt3DRender::QPickEvent *event ) override;
+  private:
+    Qgs3DMapToolMeasureLine *mMeasureLineTool = nullptr;
+};
+
+void Qgs3DMapToolMeasureLinePickHandler::handlePickOnVectorLayer( QgsVectorLayer *vlayer, QgsFeatureId id, const QVector3D &worldIntersection, Qt3DRender::QPickEvent *event )
+{
```

Comment:
looks like this function is mostly a copy of the content in onTerrainPicked() - would it be possible to merge the two to avoid code duplication?

Reply:
Done.